walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -24,41 +24,60 @@
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    Collection<TimeRange> times = new ArrayList<>();
+    // TODO: Maybe use binary tree or another data structure for availableTimes 
+    // because branches split in two?
+    // TODO: Combine the repetition of calling getAvailableTimes() below so it
+    // doesn't need to repeat without optional attendees.
+    Collection<TimeRange> availableTimes = new ArrayList<>();
     Collection<String> attendees = request.getAttendees();
-    long duration = (long) request.getDuration();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<String> allAttendees = new ArrayList<>(request.getAttendees());
+    allAttendees.addAll(optionalAttendees);
+    long duration = request.getDuration();
     if (duration > TimeRange.WHOLE_DAY.duration()){
-      return times;
+      return availableTimes;
     }
-    times.add(TimeRange.WHOLE_DAY);
+    availableTimes = getAvailableTimes(events, allAttendees, duration);
+    if (availableTimes.isEmpty()){
+      availableTimes = getAvailableTimes(events, attendees, duration);
+    }
```

Comment:
This approach is not bad, although there's some things you can do to optimize.  Some hints:
1.  Mandatory attendees always blocks the time slot.  If on your first `getAvailableTimes` call you've already determined 1pm-2pm is blocked for example, your next `getAvailableTimes` for "no optionals" shouldn't need to reevaluate that slot
2.  Optional attendees do not block the time slot, and the more of these optionals you can include in the meeting the better.  Look into ranking non-mandatory-blocked time slots based on optionals included.
3.  You don't necessarily need to know which attendee is blocked for which times.  You just need to know if the time slot is blocked by a mandatory attendee or an optional attendee.  In other words think about `__|--A--|--B--|__|--C--|--A--|__` if A and B are mandatory but C is optional, would the distinction between A and B matter?
4.  To optimize for run time, generally it'll be more efficient to do multiple separate loops than to do nested loops.  Storage is cheaper than CPU, so consider building more temporary storage (aka. caching) if it'll help avoid extra processing.  Try to think about the space/time big-O


Let me know if you need more specifics.